File: java/1370.increasing-decreasing-string.java
/*
 * @lc app=leetcode id=1370 lang=java
 *
 * [1370] Increasing Decreasing String
 */

// @lc code=start
class Solution {
    public String sortString(String s) {
        int[] count = new int[26];
        int rest = 0;
        for (int i = 0; i < s.length(); ++i) {
            ++count[s.charAt(i) - 'a'];
            ++rest;
        }
        StringBuilder res = new StringBuilder();
        while (rest > 0) {
            for (int i = 0; i < 26; ++i) {
                if (count[i] > 0) {
                    --count[i];
                    --rest;
                    res.append((char) ('a' + i));
                }
            }
            for (int i = 25; i > -1; --i) {
                if (count[i] > 0) {
                    --count[i];
                    --rest;
                    res.append((char) ('a' + i));
                }
            }
        }
        return res.toString();
    }
}
// @lc code=end

File: cpp/1370.increasing-decreasing-string.cpp
/*
 * @lc app=leetcode id=1370 lang=cpp
 *
 * [1370] Increasing Decreasing String
 */
#include <string>
#include <vector>
using namespace std;
// @lc code=start
class Solution
{
public:
    string sortString(string s)
    {
        vector<int> count(26);
        for (auto c : s)
        {
            ++count[c - 'a'];
        }
        string res;
        while (count != vector<int>(26))
        {
            for (int i = 0; i != 26; ++i)
            {
                if (count[i])
                {
                    --count[i];
                    res.push_back('a' + i);
                }
            }
            for (int i = 25; i != -1; --i)
            {
                if (count[i])
                {
                    --count[i];
                    res.push_back('a' + i);
                }
            }
        }
        return res;
    }
};
// @lc code=end
